File: matlab/tests/test_select_pre_maet.m
%% test_select_pre_maet.m — keeping a selection of a pre-MAET
%
%  Mirror of Python tests/test_select_pre_maet.py.
%
%  Standalone-runnable; appends to `results` when called from test_mpt.m.

if ~exist('results', 'var')
    results = {};
    standalone = true;
else
    standalone = false;
end

sp_p = {[60 64 67; 72 76 79], [0 1 2]};
sp_w = {[1 2 3; 1 1 1], ones(1, 3)};
sp_specs = flatSpecs(sp_p, 'r', [2 1], 'exch', [false true], ...
                     'name', {'pitch', 'onset'});
sp_pm = @() preMaet(sp_p, sp_w, sp_specs);

[p, ~, sp] = unpackPreMaet(selectPreMaet(sp_pm(), 'attributes', {'onset'}));
results(end+1, :) = {'select: attributes by name', ...
    isequal(cellfun(@(s) s.name, sp, 'UniformOutput', false), {'onset'}) ...
    && isequal(size(p{1}), [1 3])}; %#ok<*SAGROW>

[~, ~, sp] = unpackPreMaet(selectPreMaet(sp_pm(), 'attributes', [2 1]));
results(end+1, :) = {'select: attributes by index, in the order given', ...
    isequal(cellfun(@(s) s.name, sp, 'UniformOutput', false), {'onset', 'pitch'})};

[~, ~, sp] = unpackPreMaet(selectPreMaet(sp_pm(), 'attributes', [true false]));
results(end+1, :) = {'select: attributes by mask', ...
    isequal(cellfun(@(s) s.name, sp, 'UniformOutput', false), {'pitch'})};

[p, w, ~] = unpackPreMaet(selectPreMaet(sp_pm(), 'events', [1 3]));
results(end+1, :) = {'select: events by index', ...
    isequal(size(p{1}), [2 2]) && isequal(p{2}, [0 2]) ...
    && isequal(w{1}(1, :), [1 3])};

[p, ~, ~] = unpackPreMaet(selectPreMaet(sp_pm(), 'events', [false true false]));
results(end+1, :) = {'select: events by mask', isequal(p{2}, 1)};

[p, ~, sp] = unpackPreMaet(selectPreMaet(sp_pm(), ...
    'attributes', {'pitch'}, 'events', 3));
results(end+1, :) = {'select: both at once', ...
    isequal(cellfun(@(s) s.name, sp, 'UniformOutput', false), {'pitch'}) ...
    && isequal(p{1}, [67; 79])};

% A selection cannot change what an attribute means.
[~, ~, sp] = unpackPreMaet(selectPreMaet(sp_pm(), 'events', [1 2]));
results(end+1, :) = {'select: an attribute keeps its tuple size and flags', ...
    sp{1}.r == 2 && sp{1}.exch == false};

% A level's simplex coordinates are one attribute, not several, so no
% selection over attributes can take part of one.
sp_vert = simplexVertices(4);
sp_pmv = preMaet({[60 64], sp_vert(1:2, :).'}, [], ...
    flatSpecs({[60 64], zeros(3, 2)}, 'r', [1 3], 'exch', [true false], ...
              'name', {'pitch', 'voice'}));
[p, ~, sp] = unpackPreMaet(selectPreMaet(sp_pmv, 'attributes', {'voice'}));
results(end+1, :) = {'select: a multi-coordinate attribute moves whole', ...
    isequal(size(p{1}), [3 2]) && sp{1}.r == 3};

results(end+1, :) = {'select: keeping no attribute is refused', ...
    throwsErrorWithId(@() selectPreMaet(sp_pm(), 'attributes', false(1, 2)), ...
                      'selectPreMaet:emptySelection')};

results(end+1, :) = {'select: an unknown name is refused', ...
    throwsErrorWithId(@() selectPreMaet(sp_pm(), 'attributes', {'velocity'}), ...
                      'selectPreMaet:unknownName')};

results(end+1, :) = {'select: an out-of-range index is refused', ...
    throwsErrorWithId(@() selectPreMaet(sp_pm(), 'events', 5), ...
                      'selectPreMaet:outOfRange')};

results(end+1, :) = {'select: a wrong-length mask is refused', ...
    throwsErrorWithId(@() selectPreMaet(sp_pm(), 'events', [true false]), ...
                      'selectPreMaet:maskLength')};

% Allowed, and it means what it says: the event contributes nothing on
% that attribute while keeping its place.
sp_gap = preMaet({[60 NaN], [0 1]}, {[1 0], ones(1, 2)}, ...
                 flatSpecs({[60 NaN], [0 1]}, 'name', {'pitch', 'onset'}));
[p, ~, ~] = unpackPreMaet(selectPreMaet(sp_gap, 'events', 2));
results(end+1, :) = {'select: selecting may leave an event with no value', ...
    isnan(p{1}(1, 1))};

sp_out = selectPreMaet(sp_pm(), 'attributes', {'pitch'}, 'events', [1 2]);
sp_d = buildMaet(sp_out.pAttr, sp_out.wAttr, 1, 2, false, false, 0, ...
                 'verbose', false);
results(end+1, :) = {'select: the result is a pre-MAET that builds', ...
    abs(simMaet(sp_d, sp_d, 'verbose', false) - 1) < 1e-12};

if standalone
    nPass = sum([results{:, 2}]);
    fprintf('\n=== test_select_pre_maet: %d passed, %d failed (of %d) ===\n', ...
            nPass, size(results, 1) - nPass, size(results, 1));
end

File: python/tests/test_select_pre_maet.py
"""Keeping a selection of a pre-MAET's attributes and events.

Mirror of MATLAB tests/test_select_pre_maet.m.
"""

import numpy as np
import pytest

from mpt import (build_maet, flat_specs, pre_maet, select_pre_maet,
                 sim_maet, unpack_pre_maet)


def _pm():
    p = [np.array([[60.0, 64.0, 67.0], [72.0, 76.0, 79.0]]),
         np.array([[0.0, 1.0, 2.0]])]
    w = [np.array([[1.0, 2.0, 3.0], [1.0, 1.0, 1.0]]),
         np.ones((1, 3))]
    return pre_maet(p, w, flat_specs(p, r=[2, 1], exch=[False, True],
                                     name=["pitch", "onset"]))


def test_attributes_by_name():
    p, w, specs = unpack_pre_maet(select_pre_maet(_pm(), attributes=["onset"]))
    assert [s["name"] for s in specs] == ["onset"]
    assert p[0].shape == (1, 3)


def test_attributes_by_index_and_order():
    _, _, specs = unpack_pre_maet(select_pre_maet(_pm(), attributes=[1, 0]))
    assert [s["name"] for s in specs] == ["onset", "pitch"]


def test_attributes_by_mask():
    _, _, specs = unpack_pre_maet(
        select_pre_maet(_pm(), attributes=[True, False]))
    assert [s["name"] for s in specs] == ["pitch"]


def test_events_by_index():
    p, w, _ = unpack_pre_maet(select_pre_maet(_pm(), events=[0, 2]))
    assert p[0].shape == (2, 2)
    np.testing.assert_array_equal(p[1], [[0.0, 2.0]])
    np.testing.assert_array_equal(w[0][0], [1.0, 3.0])


def test_events_by_mask():
    p, _, _ = unpack_pre_maet(
        select_pre_maet(_pm(), events=[False, True, False]))
    np.testing.assert_array_equal(p[1], [[1.0]])


def test_both_at_once():
    p, _, specs = unpack_pre_maet(
        select_pre_maet(_pm(), attributes=["pitch"], events=[2]))
    assert [s["name"] for s in specs] == ["pitch"]
    np.testing.assert_array_equal(p[0], [[67.0], [79.0]])


def test_an_attribute_keeps_its_tuple_size_and_flags():
    """A selection cannot change what an attribute means."""
    _, _, specs = unpack_pre_maet(select_pre_maet(_pm(), events=[0, 1]))
    assert specs[0]["r"] == 2
    assert specs[0]["exch"] is False


def test_a_multi_coordinate_attribute_moves_whole():
    """A level's simplex coordinates are one attribute, not several, so
    no selection over attributes can take part of one."""
    coords = np.array([[0.5, -0.5], [0.2887, 0.2887], [0.2041, 0.2041]])
    p = [np.array([[60.0, 64.0]]), coords]
    pm = pre_maet(p, None, flat_specs(p, r=[1, 3], exch=[True, False],
                                      name=["pitch", "voice"]))
    out, _, specs = unpack_pre_maet(select_pre_maet(pm, attributes=["voice"]))
    assert out[0].shape == (3, 2)
    assert specs[0]["r"] == 3


def test_keeping_no_attribute_is_refused():
    with pytest.raises(ValueError, match="keeps no attribute"):
        select_pre_maet(_pm(), attributes=[])


def test_an_unknown_name_is_refused():
    with pytest.raises(KeyError, match="velocity"):
        select_pre_maet(_pm(), attributes=["velocity"])


def test_an_out_of_range_index_is_refused():
    with pytest.raises(IndexError, match="out of range"):
        select_pre_maet(_pm(), events=[5])


def test_a_wrong_length_mask_is_refused():
    with pytest.raises(ValueError, match="length 3"):
        select_pre_maet(_pm(), events=[True, False])


def test_selecting_may_leave_an_event_with_no_value():
    """Allowed, and it means what it says: the event contributes nothing
    on that attribute while keeping its place."""
    p = [np.array([[60.0, np.nan]]), np.array([[0.0, 1.0]])]
    w = [np.array([[1.0, 0.0]]), np.ones((1, 2))]
    pm = pre_maet(p, w, flat_specs(p, name=["pitch", "onset"]))
    out, _, _ = unpack_pre_maet(select_pre_maet(pm, events=[1]))
    assert np.isnan(out[0][0, 0])


def test_the_result_is_a_pre_maet_that_builds():
    pm = select_pre_maet(_pm(), attributes=["pitch"], events=[0, 1])
    p, w, specs = unpack_pre_maet(pm)
    dens = build_maet(p, w, [1.0], [2], [False], [False], [0.0],
                      verbose=False)
    assert sim_maet(dens, dens, verbose=False) == pytest.approx(1.0)
